File: lib/models/product.model.dart
class Product {
  final String id;
  final String name;
  final String description;
  final double price;
  final String categoryId;
  final int stock;
  final String status;
  final String? image; // Cho phép null nếu server có thể không trả về
  final bool featured;
  final double rating;
  final String slug;

  Product({
    required this.id,
    required this.name,
    required this.description,
    required this.price,
    required this.categoryId,
    required this.stock,
    required this.status,
    this.image,
    required this.featured,
    required this.rating,
    required this.slug,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      price: (json['price'] is int)
          ? (json['price'] as int).toDouble()
          : (json['price'] as double? ?? 0.0),
      categoryId: json['category']?['_id'] ?? '',
      stock: json['stock'] ?? 0,
      status: json['status'] ?? '',
      image: json['image'],
      featured: json['featured'] ?? false,
      rating: (json['rating'] is int)
          ? (json['rating'] as int).toDouble()
          : (json['rating'] as double? ?? 0.0),
      slug: json['slug'] ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        '_id': id, // Nếu bạn cần gửi lại dữ liệu này tới server
        'name': name,
        'description': description,
        'price': price,
        'categoryId': categoryId, // Gửi lại dưới dạng ID của danh mục
        'stock': stock,
        'status': status,
        'image': image,
        'featured': featured,
        'rating': rating,
        'slug': slug,
      };
}

File: src/v1/models/product.model.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const slugify = require('slugify');

const COLLECTION_NAME = 'Products';
const DocumentName = 'Product';

const productSchema = new Schema({
    name: { type: String, required: true },
    description: { type: String, required: true },
    price: { type: Number, required: true },
    category: { type: Schema.Types.ObjectId, ref: 'Category' },
    stock: { type: Number, required: true },
    status: { type: String, enum: ['active', 'inactive'], default: 'active' },
    image: { type: String },
    // images: [{ type: String }], // Mảng URL hình ảnh
    featured: { type: Boolean, default: false },
    rating: { type: Number, default: 0, min: 0, max: 5, set: v => Math.round(v * 10) / 10},
    slug: { type: String, unique: true },
},{
    timestamps: true,
    collection: COLLECTION_NAME
});

// Middleware để tự động tạo slug từ tên sản phẩm
productSchema.pre('save', function(next) {
    this.slug = slugify(this.name, { lower: true });
    next();
});

// Tạo index cho tìm kiếm hiệu quả
productSchema.index({ name: 'text', description: 'text' });

module.exports = mongoose.model(DocumentName, productSchema, COLLECTION_NAME);
